Update install-a-docker-delegate.md (#347)
harness/developer-hub: docs/platform/2_Delegates/delegate-install-docker/install-a-docker-delegate.md

helpdocs_is_published: true
---

> Harness Delegate is a worker process you run in your deployment target environment, such as your local network, VPC, or cluster. The delegate connects artifact, infrastructure, collaboration, verification and other providers with Harness Manager.

Author: Can't add a comma there. It wouldn't be grammatical since it separates "verification" from "and other providers".

### Review: System Requirements

Docker Delegate has the following system requirements:

* Default 0.5 CPU.

> * Default 768MB RAM. There is a cap of 768MB per delegate, but when the delegate is updating there might be two delegates running. Hence, the minimum is 1.5 GB. Ensure that you provide the minimum memory for the delegate and enough memory for the host/node system.

Author: There are a lot of editorial changes I would like to make to the document but the main goal of the changes was to fix problems that occurred in the migration so I make only limited copy edits.

Typically I avoid the use of "its" and I would write around some of the problems you've noticed but wouldn't be comfortable assuming that when two delegates are running, one of which is being updates, that the tasks of the first delegate are performed by the second and that we consider the second temporary. It may be implied. But that's the kind of thing that I'd probably want to have reviewed.

> * Default 768MB RAM. There is a cap of 768MB per delegate, but when the delegate is updating there might be two delegates running. Hence, the minimum is 1.5 GB. Ensure that you provide the minimum memory for the delegate and enough memory for the host/node system.

Author: FYI, I avoid the use of the "its" because it can be vague.

### Step 1: Download Docker Delegate

> Docker Delegate can be installed at the Harness account, Organization, or Project level.

Author: The word "scope" doesn't match the current preposition "at". Programming elements are scoped or given a specific scope. I think the writer was talking about account permissions in a hierarchy of user permissions. Scope is often associated with namespace. I think it's fine as written and must assume that it was technically reviewed when it was written.

> You can install Docker Delegate while setting up a connector or independent of another process.

Reviewer: I'm not sure that the clauses joined by the "or" are similar (or parallel) in nature. Also, "as part of" might have been better than "while"? The "while" does not suggest that installing Docker Delegate is part of the connector configuration workflow (or the workflow for any entity). They might still be independent of each other; just performed in parallel. How about "Various workflows for configuring Harness entities, such as those for configuring connectors, include a step for installing Docker Delegate. However, you can also install Docker Delegate beforehand and then reference them in those workflows." (I'm going wild here but please consider this sample text. Also, I felt this is a classic case of what the product provides vs. what the user can do, so I modified the wording accordingly.)

Reviewer: Subsequent text suggests that the connector workflow is the only workflow that includes this installation option. If my recommendation is good, please modify it as needed. Thanks.

> After you select **New Delegate** on the Delegate page or while setting up a connector, the delegate selection page appears.

Reviewer: ```suggestion
After you select **New Delegate** on the Delegate page, or when you are setting up a connector, the delegate selection page appears.
```

![](./static/install-a-docker-delegate-00.png)

Click **Docker** and then click **Continue**.

### Step 2: Name and tag the delegate

Enter a name for the delegate. This name will identify the delegate in Harness Manager.

> **Do not run delegates with the same name in different clusters.** See [Troubleshooting](../../../troubleshooting/troubleshooting-nextgen.md).Add Tags to the Delegate. By default, Harness adds a tag using the name you enter. To add more, type them in and press **Enter**.

Reviewer: ```suggestion
**Do not run delegates with the same name in different clusters.** See [Troubleshooting](../../../troubleshooting/troubleshooting-nextgen.md). Add tags to the delegate. By default, Harness adds a tag that uses the name you enter. To add more, type them in and press **Enter**.
```

Reviewer: * "Type them in" > "type them in <some box>" (just thinking why we wrote "in". Might be colloquial too, in which case you don't need to enter a box name or something else. 
*  **Do not run delegates with the same name in different clusters.** > Do we need bold text for emphasis? Or can we use a note instead? Also, does that mean we can run delegates with the same name on the same cluster? Or should we not have delegates with the same name at all?

These tags are useful for selecting the delegate when creating a connector.

![](./static/install-a-docker-delegate-01.png)

Click **Continue**.

### Step 3: Run the Docker Delegate script

> If you system already has a delegate image, Harness doesn't pull the latest image when you run `docker-compose`. Simply run `docker pull harness/delegate` to pull the latest. Now you can see the YAML file for the delegate:

Reviewer: "If your"

![](./static/install-a-docker-delegate-02.png)
Click **Download YAML file** and copy the Docker compose file to a machine where you have Docker installed.

Run the following command to install the delegate in Docker:

```
docker-compose -f docker-compose.yaml up -d
```

> The delegate installs. Type `docker ps` to see the container:

Reviewer: "is installed"?

For an overview of verification, see [Delegate Registration and Verification](../delegate-guide/delegate-registration.md).

> In the delegate wizard, click **Verify** and Harness will verify that it is receiving heartbeats from the delegate.

Reviewer: "for Harness to verify"
Alternatively "click Verify to verify that Harness is receiving ..." (since the user will want to verify, not Harness itself)

Your delegate is installed.

The registered delegate appears in the delegate list.

![](./static/install-a-docker-delegate-03

Note the **Connected** status. If there is a connectivity error, you will see **Not Connected**. If there's an error, ensure the Docker host can connect to `https://app.harness.io`.

> That's it. The delegate is installed and registered and connected.

Reviewer: ```suggestion
That's it. The delegate is installed, registered, and connected.
```

### Harness Docker Delegate Environment Variables

The following table lists each of the environment variables in the application manifest of the Docker Delegate.

| **Name** | **Description** | **Example** |
| --- | --- | --- |

> | `ACCOUNT_ID` | The Harness account ID for the account where this Delegate will attempt to register. | ```- ACCOUNT_ID=H5W8i2828282828Xg```|

Reviewer: ```suggestion
| `ACCOUNT_ID` | The ID of the Harness account with which the Delegate attempts to register. | ```- ACCOUNT_ID=H5W8i2828282828Xg```|
```

| `DELEGATE_TOKEN` | The Harness account token used to register the Delegate. | ```- DELEGATE_TOKEN=d229ee88bf7bbxxxx6ea```|
| `MANAGER_HOST_AND_PORT` | The Harness SaaS manager URL. `https` indicates port 443. |```- MANAGER_HOST_AND_PORT=https://app.harness.io```|
| `WATCHER_STORAGE_URL` | The URL for the Watcher versions. See [Delegate Installation Overview](../delegate-installation-overview.md). | ```- WATCHER_STORAGE_URL=https://app.harness.io/public/prod/premium/watchers```|
| `WATCHER_CHECK_LOCATION` | The Delegate version location for the Watcher to check for. | ```- name: WATCHER_CHECK_LOCATION  value: current.version```|
| `REMOTE_WATCHER_URL_CDN` | The CDN URL for Watcher builds. | ```- name: REMOTE_WATCHER_URL_CDN  value: https://app.harness.io/public/shared/watchers/builds```|

> | `DELEGATE_STORAGE_URL` | The URL where published delegate jars are stored. | ```- name: DELEGATE_STORAGE_URL  value: https://app.harness.io```|

Reviewer: ```suggestion
| `DELEGATE_STORAGE_URL` | The URL at which published delegate jars are stored. | ```- name: DELEGATE_STORAGE_URL  value: https://app.harness.io```|
```

| `DELEGATE_CHECK_LOCATION` | The storage location hosting the published delegate versions. | ```- name: DELEGATE_CHECK_LOCATION  value: delegateprod.txt```|
| `DEPLOY_MODE` | Deployment mode: Kubernetes, Docker, and so on | ```- name: DEPLOY_MODE  value: DOCKER```|

> | `DELEGATE_NAME` | The name of the delegate. This is the name that appears in Harness when the delegate is registered. You can automate delegate creation by omitting the name, and then have a script copying the delegate YAML file and add a unique name to `value` for each new delegate you want to register. See [Automate Delegate Installation](../delegate-guide/automate-delegate-installation.md). | ```- name: DELEGATE_NAME  value: qa```|

Reviewer: The "you can automate" sentence is long. Can we split it?

> | `NEXT_GEN` | Indicates that this delegate will register in Harness NextGen(`true`) or FirstGen(`false`). | ```- name: NEXT_GEN  value: "true"```|

Reviewer: ```suggestion
| `NEXT_GEN` | If `true, the delegate registers with Harness NextGen. If `false`, the delegate registers with Harness FirstGen. | ```- name: NEXT_GEN  value: "true"```|
```

> | `DELEGATE_DESCRIPTION` | The description added to the delegate in Harness Manager or YAML before registering. It appears in the delegate details page in the Harness Manager. | ```- name: DELEGATE_DESCRIPTION  value: ""```|

Reviewer: "or in the delegate YAML"?

| `DELEGATE_TYPE` | The type of delegate. | ```- name: DELEGATE_TYPE  value: "DOCKER"```|
| `DELEGATE_TAGS` | The tags added to the delegate in Harness Manager or YAML before registering. Tags are generated by Harness using the delegate name but you can also add your own tags. Tags appear in the delegate details page in Harness Manager. See [Tags Reference](../../20_References/tags-reference.md) and [Select Delegates with Tags](../delegate-guide/select-delegates-with-selectors.md). | ```- name: DELEGATE_TAGS  value: ""```|

> | `DELEGATE_TASK_LIMIT` | The maximum number of tasks the delegate can perform at once. All of the operations performed by the delegate are categorized as different types of tasks. | ```- name: DELEGATE_TASK_LIMIT  value: "50"```|

Reviewer: at once > concurrently?

| `DELEGATE_ORG_IDENTIFIER` | The Harness Organization [Identifier](../../20_References/entity-identifier-reference.md) where the delegate will register. Delegates at the account level do not have a value for this variable. | ```- name: DELEGATE_ORG_IDENTIFIER  value: "engg"```|

> | `DELEGATE_PROJECT_IDENTIFIER` | The Harness project [identifier](../../20_References/entity-identifier-reference.md) where the delegate will register. Delegates at the account or Org-level do not have a value for this variable. | ```-name: DELEGATE_PROJECT_IDENTIFIER  value: "myproject"```|

Reviewer: ```suggestion
| `DELEGATE_PROJECT_IDENTIFIER` | The Harness project [identifier](../../20_References/entity-identifier-reference.md) with which the delegate registers. Delegates at the account or org level do not have a value for this variable. | ```-name: DELEGATE_PROJECT_IDENTIFIER  value: "myproject"```|
```

Reviewer: Also, prefer "organization" to "org".

| `PROXY_MANAGER` | Indicates whether to use the Harness Manager or a proxy. If you use `true` (the default) you are indicating that you proxy outbound traffic to Harness. | ```- PROXY_MANAGER=true```|

> | `INIT_SCRIPT` | You can run scripts on the delegate using `INIT_SCRIPT`. For example, if you wanted to install software on the delegate pod, you can enter the script in `INIT_SCRIPT` and then apply the delegate YAML. A multiline script must follow the YAML `spec` for [literal scalar style](https://yaml.org/spec/1.2-old/spec.html#id2795688). See [Run Scripts on Delegates](../delegate-guide/run-scripts-on-delegates.md). | ```- INIT_SCRIPT=  echo hello world!```|

Reviewer: ```suggestion
| `INIT_SCRIPT` | You can run scripts on the delegate by using `INIT_SCRIPT`. For example, if you want to install software on the delegate pod, you can enter the script in `INIT_SCRIPT` and then apply the delegate YAML. A multiline script must follow the YAML `spec` for [literal scalar style](https://yaml.org/spec/1.2-old/spec.html#id2795688). See [Run Scripts on Delegates](../delegate-guide/run-scripts-on-delegates.md). | ```- INIT_SCRIPT=  echo hello world!```|
```

Reviewer: * Good to expand on what `INIT_SCRIPT` is. For example, "the `INIT_SCRIPT` environment variable" (for example)
* "you can enter the script" > should that be "you can enter the path to the script"? (ignore if my question is invalid)

> | `USE_CDN` | Makes the delegate use a CDN for new versions. | ```- name: USE_CDN  value: "true"```|

Reviewer: Should we expand CDN?

| `CDN_URL` | The CDN URL for delegate versions. | ```- name: CDN_URL  value: https://app.harness.io```|

> | `VERSION_CHECK_DISABLED` | By default, the delegate always checks for new versions (via the Watcher). | ```- name: VERSION_CHECK_DISABLED  value: "false"```|

Reviewer: via > through?
